packages/opencensus-propagation-tracecontext/src/tracecontext-format.ts

```diff
@@ -0,0 +1,142 @@
+/**
+ * Copyright 2018, OpenCensus Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {HeaderGetter, HeaderSetter, Propagation, SpanContext} from '@opencensus/core';
+import * as crypto from 'crypto';
+
+import {isValidOption, isValidSpanId, isValidTraceId, isValidVersion} from './validators';
+
+// Header names
+const TRACE_PARENT = 'traceparent';
+const TRACE_STATE = 'tracestate';
+
+// Option flags
+const REQUESTED_FLAG = 0x1;
+const RECORDED_FLAG = 0x2;
+
+const DEFAULT_OPTIONS = 0x0;
+
+/**
+ * Propagates span context through Trace Context format propagation.
+ *
+ * Based on the Trace Context specification:
+ * https://w3c.github.io/distributed-tracing/report-trace-context.html
+ *
+ * Known Limitations:
+ * - Multiple `tracestate` headers are merged into a single `tracestate`
+ *   during injection. This might exceed the 512 character rule as defined
+ *   in section 2.3.2.
+ */
+export class TraceContextFormat implements Propagation {
+  /**
+   * Gets the trace context from a request headers. If there is no trace context
+   * in the headers, or if the parsed `traceId` or `spanId` is invalid, an empty
+   * context is returned.
+   * @param getter
+   */
+  extract(getter: HeaderGetter): SpanContext {
+    if (getter) {
+      // Construct empty span context that we will fill
+      const spanContext: SpanContext = {
+        traceId: undefined,
+        spanId: undefined,
+        options: DEFAULT_OPTIONS,
+        traceState: undefined
+      };
+
+      let traceState = getter.getHeader(TRACE_STATE);
+      if (Array.isArray(traceState)) {
+        // If more than one `tracestate` header is found, we merge them into a
+        // single header.
+        traceState = traceState.join(',');
+      }
+      spanContext.traceState =
+          typeof traceState === 'string' ? traceState : undefined;
+
+      // Read headers
+      let traceParent = getter.getHeader(TRACE_PARENT);
+      if (Array.isArray(traceParent)) {
+        traceParent = traceParent[0];
+      }
+
+      // Parse TraceParent into version, traceId, spanId, and option flags. All
+      // parts of the header should be present or it is considered invalid.
+      const parts = traceParent ? traceParent.split('-') : [];
+      if (parts.length === 4) {
```

Comment:
Would it make more sense to return `null` if this condition (or subsequent conditions) isn't true?

Reply:
I think (1) is a better choice, even if it involves further modifying the Tracer. From the spec:

> Notably, the tracestate field is unreliant on data in the traceparent.

So while it might not happen typically, a `tracestate` isn't necessarily invalid if sent in solation.